chalice/local/internal.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+"""
+This is intended only for `pytest-chalice`.
+
+https://github.com/studio3104/pytest-chalice
+"""
+
+import json
+import re
+
+from logging import getLogger
+from typing import Any, Callable, Dict, Optional  # noqa
+
+from chalice import Chalice  # noqa
+from chalice.config import Config
+from chalice.local import LocalGateway
+from chalice.local import EventType, HeaderType  # noqa
+
+logger = getLogger(__name__)
+
+# Python 3.4 or older don't have JSONDecodeError within json module
+try:
+    JSONDecodeError = json.JSONDecodeError  # type: ignore
+except AttributeError:
+    JSONDecodeError = ValueError  # type: ignore
+
+
+class InternalLocalGateway(LocalGateway):
+    def __init__(self, *args, **kwargs):
+        # type: (Any, Any) -> None
+        self.__custom_context = {}  # type: Dict
+        super(InternalLocalGateway, self).__init__(*args, **kwargs)
+
+    @property
+    def custom_context(self):
+        # type: () -> Dict[str, Any]
+        return self.__custom_context
+
+    @custom_context.setter
+    def custom_context(self, context):
+        # type: (Dict[str, Any]) -> None
+        self.__custom_context = context
+
+    def _generate_lambda_event(self, method, path, headers, body):
+        # type: (str, str, HeaderType, Optional[str])-> EventType
+        event = super(
+            InternalLocalGateway, self)._generate_lambda_event(
+                method, path, headers, body)
+        event['requestContext'].update(self.custom_context)
+        return event
+
+
+UPPERCASE_PATTERN = re.compile('([A-Z])')
+
+
+class ResponseHandler(object):
```

Comment:
I think we should have an abstraction on top of the response dictionary returned.  A user doesn't return this dictionary directly, Chalice automatically creates this value so I'm concerned in exposing this directly as the API you assert against.  This means that the dictionary occasionally changes when API Gateway adds support for more keys.  For example, we now have support for `multiValueHeaders`.

If we have our own response class, we could add these changes in a backwards compatible way.

What do you all think?   

cc @stealthycoin

Reply:
I may not be able to understand what you suggest right now.
You mean the class must not generate the values dynamically within the instance but map specific keys for each items, right?
Sorry to bother you but could you please show me a concrete example if not so...?